beacon-chain/state/stategen/epoch_boundary_state.go

```diff
@@ -0,0 +1,156 @@
+type rootStateInfo struct {
+	root  [32]byte
+	state *stateTrie.BeaconState
+}
+
+// slotRootInfo specifies the slot root info in the epoch boundary state cache.
+type slotRootInfo struct {
+	root [32]byte
+	slot uint64
+}
+
+// rootKeyFn takes the string representation of the block root to be used as key
+// to retrieve epoch boundary state.
+func rootKeyFn(obj interface{}) (string, error) {
+	s, ok := obj.(*rootStateInfo)
+	if !ok {
+		return "", errNotRootStateInfo
+	}
+	return string(s.root[:]), nil
+}
+
+// slotKeyFn takes the string representation of the slot to be used as key
+// to retrieve root.
+func slotKeyFn(obj interface{}) (string, error) {
+	s, ok := obj.(*slotRootInfo)
+	if !ok {
+		return "", errNotSlotRootInfo
+	}
+	return slotToString(s.slot), nil
+}
+
+// epochBoundaryState struct with two queues by looking up by slot or root.
+type epochBoundaryState struct {
+	rootStateCache *cache.FIFO
+	slotRootCache  *cache.FIFO
+	lock           sync.RWMutex
+}
+
+// newBoundaryStateCache creates a new block epochBoundaryState for storing/accessing epoch boundary states from
+// memory.
+func newBoundaryStateCache() *epochBoundaryState {
+	return &epochBoundaryState{
+		rootStateCache: cache.NewFIFO(rootKeyFn),
+		slotRootCache:  cache.NewFIFO(slotKeyFn),
+	}
+}
+
+// get epoch boundary state by its block root. Returns copied state in state info object if exists. Otherwise returns nil.
+func (e *epochBoundaryState) getByRoot(r [32]byte) (*rootStateInfo, bool, error) {
+	e.lock.RLock()
+	defer e.lock.RUnlock()
+
+	obj, exists, err := e.rootStateCache.GetByKey(string(r[:]))
+	if err != nil {
+		return nil, false, err
+	}
+	if !exists {
+		return nil, false, nil
+	}
+	s, ok := obj.(*rootStateInfo)
+	if !ok {
+		return nil, false, errNotRootStateInfo
+	}
+
+	return &rootStateInfo{
+		root:  r,
+		state: s.state.Copy(),
+	}, true, nil
+}
+
+// get epoch boundary state by its slot. Returns copied state in state info object if exists. Otherwise returns nil.
+func (e *epochBoundaryState) getBySlot(s uint64) (*rootStateInfo, bool, error) {
+	e.lock.RLock()
+	defer e.lock.RUnlock()
+
+	obj, exists, err := e.slotRootCache.GetByKey(slotToString(s))
+	if err != nil {
+		return nil, false, err
+	}
+	if !exists {
+		return nil, false, nil
+	}
+	info, ok := obj.(*slotRootInfo)
+	if !ok {
+		return nil, false, errNotSlotRootInfo
+	}
+
+	return e.getByRoot(info.root)
+}
+
+// put adds a state to the epoch boundary state cache. This method also trims the
+// least recently added state info if the cache size has reached the max cache
+// size limit.
+func (e *epochBoundaryState) put(r [32]byte, s *stateTrie.BeaconState) error {
+	e.lock.Lock()
+
+	if err := e.rootStateCache.AddIfNotPresent(&rootStateInfo{
+		root:  r,
+		state: s.Copy(),
+	}); err != nil {
+		return err
+	}
+	if err := e.slotRootCache.AddIfNotPresent(&slotRootInfo{
+		root: r,
+		slot: s.Slot(),
+	}); err != nil {
+		return err
+	}
+
+	e.lock.Unlock()
```

Comment:
Why do you return the lock before the trim operation? Shouldn't you hold it until the end of this method?

Reply:
Fixed in #6486 Thanks!

https://github.com/prysmaticlabs/prysm/pull/6486/files#diff-a6935ba2179f800c8a459f1c95ac6a26R116